[FRAME-156]: Add Events

tarecord opened the pull request:
This adds the ability to trigger specific events with a `do_action()` so additional data can be collected from sites when users perform specific actions.

ChrisMKindred reviewed (approved):
This all looks good. Should we look into using version/tag branches so you don't have to use the TBD in the since param?  I am concerned that merging directly to Main without a version branch could result in that `TBD` getting left in before a deployment.

tarecord commented:
Yeah, that's a good point and I thought the same thing.

Now that this has been released, should we move to a more strict gitflow workflow:
- Set up a new `develop` branch as the default
  - New development is merged directly into this branch
- Use specific `release` branches to prep releases
  - Once we decide we're ready for a new release, these are used to prep and merge into `main`
  - This would be where we update all `@since TBD` instances to the current release

![03 Release branches](https://github.com/stellarwp/telemetry/assets/6947218/2358f987-ff69-4da2-88bc-523a04e977ba)

ChrisMKindred commented:
I recommend creating a `contributing.md` file with the details so that it is well documented.  [GitHub will highlight this file](https://docs.github.com/en/communities/setting-up-your-project-for-healthy-contributions/setting-guidelines-for-repository-contributors) for you when people are making new contributions making it clear what the expectations are.

tarecord commented:
Hey @ChrisMKindred, [here's a link to the subset of most recent changes](https://github.com/stellarwp/telemetry/pull/77/files/f08f2a71556a48634c43d4cf9a24236f9093b6c6..c6548640ddbda1c91464a099ba2dd76a2fa74057) that should make it a little easier to see what I did.

Mainly two things:
- Moved event-related methods from `Telemetry.php` to `Event.php`
- Updated the `Telemetry::send()` method to check if the site is currently opted-in before sending

Let me know if you have any questions or suggestions.

ChrisMKindred commented on tests/wpunit/EventTest.php:
Thank You!

ChrisMKindred commented on tests/wpunit/EventTest.php:
```diff
@@ -0,0 +1,75 @@
+	 * @test
+	 */
+	public function should_send_event(): void {
+		$mock_response = [
+			'headers'       => [],
+			'body'          => json_encode( // phpcs:ignore WordPress.WP.AlternativeFunctions.json_encode_json_encode
+				[
+					'status'  => 'true',
+					'message' => 'success',
+				]
+			),
+			'response'      => [
+				'code'    => false,
+				'message' => false,
+			],
+			'cookies'       => [],
+			'http_response' => null,
+		];
+		$call_url      = null;
+		$call_args     = null;
+		$this->set_fn_return(
+			'wp_remote_post',
+			static function ( string $url, array $args ) use ( $mock_response, &$call_url, &$call_args ) {
+				$call_url  = $url;
+				$call_args = $args;
+
+				return $mock_response;
+			},
+			true
+		);
```
I could see this being used by many of our tests. Should it be refactored into a function added at setup so you don't have to add it to each test?

tarecord replied:
I wish... I spent some time working through that and there isn't a good way to store the `call_url` and `call_args` because the function passed to `$this->set_fn_return` runs statically and can't access the object running the test.

Unless you know of a way to access those, I'm going to leave this within the test.

ChrisMKindred reviewed (approved):
Looks good and is fine to merge. Could we add the mocking as a setup function?  It is repetitive for us to create the array and function mock in each test.
